### PR TITLE
Milestone 0.5 Updates

### LINKED ISSUE #1: Hide Visibility and Date Settings in the Post Box
post-box

The _Visibility_ and _Publish_ settings on the Edit Redirect screen do not make sense in this context, so let's remove them.

### PR DESCRIPTION
**Pull request by roborourke**
Aware it's not best practice but the changes were all pretty minor for this release.

Fixes #1 #6 #11 #12 #13 #14 #15 #16 #21 #26

### CODE REVIEW

File: assets/admin.css [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#post-body-content,
+#submitdiv .misc-pub-visibility,
+#submitdiv .misc-pub-curtime { display: none; }
```
Comment (fklein-lu) [inline assets/admin.css]: This is not according to coding standards, see https://engineering.hmn.md/standards/style/css/#basic-styles

File: changelog.txt [hunk 2]
```diff
@@ -1,5 +1,18 @@
 == Changelog ==
 
+= 0.5.0 =
+* Fix redirect CPT to be private
+* Add filters to modify the contents of the HTTP Status Code drop down enhancement good first issue
```
Comment (fklein-lu) [inline changelog.txt]: Not really sure what "good first issue" has to do here.

File: changelog.txt [hunk 3]
```diff
@@ -1,5 +1,18 @@
...
+* Sanitise full from URL to relative path
+* Fix redirects with query arguments bug
```
Comment (fklein-lu) [inline changelog.txt]: What was the bug?

File: includes/admin.php [hunk 4]
```diff
@@ -16,8 +16,11 @@
  */
 function setup() {
 	add_action( 'add_meta_boxes', __NAMESPACE__ . '\\add_meta_box' );
-	add_action( 'save_post', __NAMESPACE__ . '\\handle_redirect_saving', 13 );
+	add_action( 'admin_enqueue_scripts', __NAMESPACE__ . '\\enqueue_scripts' );
+	add_action( 'manage_' . Redirects_Post_Type\SLUG . '_posts_custom_column', __NAMESPACE__ . '\\posts_columns_content', 10, 2 );
+	add_filter( 'manage_' . Redirects_Post_Type\SLUG . '_posts_columns', __NAMESPACE__ . '\\filter_posts_columns' );
 	add_action( 'plugins_loaded', __NAMESPACE__ . '\\load_plugin_textdomain' );
+	add_action( 'save_post', __NAMESPACE__ . '\\handle_redirect_saving', 13 );
```
Comment (fklein-lu) [inline includes/admin.php]: Why is this specific priority chosen over the default one? A comment would be appropriate.

File: includes/admin.php [hunk 5]
```diff
@@ -34,6 +37,51 @@ function add_meta_box() {
...
+ * @return array
+ */
+function filter_posts_columns( array $columns ) : array {
+	$columns = [
+		'title' => __( 'From', 'hm-redirects' ),
+		'to' => __( 'To', 'hm-redirects' ),
+		'status' => __( 'Status', 'hm-redirects' ),
+		'date' => __( 'Date' ),
+	];
+	return $columns;
```
Comment (fklein-lu) [inline includes/admin.php]: Blank line before returns.

File: includes/admin.php [hunk 6]
```diff
@@ -34,6 +37,51 @@ function add_meta_box() {
...
+ * Output for the custom admin columns.
+ *
+ * @param string $column Current column ID.
+ * @param int    $post_id Current post ID.
+ */
+function posts_columns_content( string $column, int $post_id ) {
+	$post = get_post( $post_id );
+
+	if ( $column === 'to' ) {
+		echo esc_html( sanitize_text_field( $post->post_excerpt ) );
```
Comment (fklein-lu) [inline includes/admin.php]: By using `sanitize_text_field()` you run into the same issue as was present in the matching of non ASCII URLs bug. Special characters get octet encoded, and these get stripped during sanitisation.

Beyond this, we are outputting data, so there should be no sanitisation applied.

File: includes/admin.php [hunk 7]
```diff
@@ -50,17 +98,41 @@ function output_meta_box( WP_Post $post ) {
 		404 => 'Not Found',
 	];
 
-	$status_code = ! empty( $post->post_content_filtered ) ? $post->post_content_filtered : 302;
+	/**
+	 * Filter available status codes.
+	 *
+	 * The array keys are the status code and the values are the labels.
+	 *
+	 * Use this filter to modify the labels or remove options.
```
Comment (fklein-lu) [inline includes/admin.php]: I don't think this explanation is needed, but could be left in.

File: includes/admin.php [hunk 8]
```diff
@@ -50,17 +98,41 @@ function output_meta_box( WP_Post $post ) {
 		404 => 'Not Found',
 	];
 
-	$status_code = ! empty( $post->post_content_filtered ) ? $post->post_content_filtered : 302;
+	/**
+	 * Filter available status codes.
+	 *
+	 * The array keys are the status code and the values are the labels.
```
Comment (fklein-lu) [inline includes/admin.php]: This needs to go into the `@param` description.

File: includes/admin.php [hunk 9]
```diff
@@ -50,17 +98,41 @@ function output_meta_box( WP_Post $post ) {
...
+	 * Use this filter to modify the labels or remove options.
+	 *
+	 * @param array $status_code_labels Array of status codes and labels.
+	 */
+	$status_code_labels = apply_filters( 'hm_redirects_status_codes', $status_code_labels );
+
+	/**
+	 * Filter the default selected status code.
+	 *
+	 * Defaults to 302.
```
Comment (fklein-lu) [inline includes/admin.php]: This needs to go into the `@param` description.

File: includes/handle-redirects.php [hunk 10]
```diff
@@ -32,10 +32,6 @@ function maybe_do_redirect() {
 
 	$path = untrailingslashit( str_replace( wp_parse_url( home_url(), PHP_URL_PATH ), '', sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ) ) );
 
-	if ( ! empty( $_SERVER['QUERY_STRING'] ) ) {
```
Comment (fklein-lu) [inline includes/handle-redirects.php]: So we just ignore this? This seems like the wrong fix to me.

File: includes/utilities.php [hunk 11]
```diff
@@ -127,6 +127,26 @@ function insert_redirect( $from, $to, $status_code, $post_id = 0 ) {
 	// Stop loops.
 	remove_action( 'save_post', 'HM\Redirects\\Admin_UI\\handle_redirect_saving', 13 );
 
+	/**
+	 * Filter the from URL value before saving to the database.
```
Comment (fklein-lu) [inline includes/utilities.php]: Needs documentation for params.

File: includes/utilities.php [hunk 12]
```diff
@@ -127,6 +127,26 @@ function insert_redirect( $from, $to, $status_code, $post_id = 0 ) {
...
+	 *
+	 * @param string $from
+	 * @param string $to
+	 * @param int $status_code
+	 * @param int $post_id
+	 */
+	$from = apply_filters( 'hm_redirects_pre_save_from_url', $from, $to, $status_code, $post_id );
+
+	/**
+	 * Filter the from URL value before saving to the database.
```
Comment (fklein-lu) [inline includes/utilities.php]: Wrong comment, plus needs documentation.

File: changelog.txt (hunk 2 above)
Comment (roborourke) [reply changelog.txt]: copy paste error - those are the github labels, I forgot to remove them

File: changelog.txt (hunk 3 above)
Comment (roborourke) [reply changelog.txt]: I'll add issue numbers - essentially it was doubling up on the query string eg. `/?this=that?this=that`

File: includes/handle-redirects.php (hunk 10 above)
Comment (roborourke) [reply includes/handle-redirects.php]: See issue #15 - the query string is already present via `REQUEST_URI` so this code just adds it again. Why not just remove this?

File: includes/admin.php (hunk 4 above)
Comment (roborourke) [reply includes/admin.php]: It's marked as changed because I put them in alphabetical order - the priority was already set to that, not sure why either.

**Comment by roborourke** [comment]
Cheers @fklein-lu - updated the code and added and some replies to specific things.

File: includes/handle-redirects.php (hunk 10 above)
Comment (roborourke) [reply includes/handle-redirects.php]: Can come back to this in the 1.0 milestone if you feel there's a more correct approach.

**Pull request merged**